frontend/src/hooks/useScreenMediaQuery.ts

```diff
@@ -0,0 +1,50 @@
+import { useEffect, useMemo, useState } from 'react';
+import SCREEN_BREAKPOINT from '../constants/screenBreakpoints';
+
+type ScreenBreakPoint = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
+
+type Screen = Record<ScreenBreakPoint, boolean>;
+
+/**
+ * React 컴포넌트에서 스크린 사이즈에 따라 실행해야 할 코드가 다를 때
+ * 이 훅을 사용하여 분기 처리를 할 수 있다.
+ *
+ * @example
+ * const screen = useScreenMediaQuery();
+ * if (screen.sm) { // 모바일 디바이스에서 처리할 로직
+ *   // ...
+ * }
+ * if (screen.xl) {
+ *   // ...
+ * }
+ */
+const useScreenMediaQuery = (): Screen => {
+  const mediaQueries = useMemo(
+    () =>
+      Object.fromEntries(
+        Object.entries(SCREEN_BREAKPOINT).map(([name, breakpoint]) => [
+          name,
+          window.matchMedia(`(max-width: ${breakpoint}px)`),
+        ])
+      ) as Record<ScreenBreakPoint, MediaQueryList>,
+    []
+  );
+
+  const [screen, setScreen] = useState<Screen>(
+    Object.fromEntries(
+      Object.entries(mediaQueries).map(([name, mediaQuery]) => [name, mediaQuery.matches])
+    ) as Screen
+  );
+
+  useEffect(() => {
+    Object.entries(mediaQueries).map(([name, mediaQuery]) => {
+      mediaQuery.addEventListener('change', ({ matches }) => {
+        setScreen((screen) => ({ ...screen, [name]: matches }));
+      });
+    });
+  }, []);
+
+  return screen;
+};
```

Comment:
```suggestion
const useScreenMediaQuery = (): ScreenBreakPoint => {
  const [screen, setScreen] = useState<ScreenBreakPoint>(calculateWindowSize(window.innerWidth));

  const handleResizeScreen = useCallback(() => {
    const newScreen = calculateWindowSize(window.innerWidth);
    if (newScreen !== screen) {
      setScreen(newScreen);
    }
  }, [screen]);

  useEffect(() => {
    window.addEventListener('resize', handleResizeScreen);
    return () => window.removeEventListener('resize', handleResizeScreen);
  }, [handleResizeScreen]);

  return screen;
};

const calculateWindowSize = (width: number): ScreenBreakPoint => {
  if (width > 1280) {
    return 'xl';
  } else if (width > 1024) {
    return 'lg';
  } else if (width > 768) {
    return 'md';
  } else if (width > 640) {
    return 'sm';
  } else {
    return 'xs';
  }
};
```
저는 이런 느낌으로 사용하는게 좀 더 직관적인 것 같아요. 근데 사실 이 훅 내부를 뜯어볼 일은 없을 것 같아서 가독성을 굳이 챙길 필요는 없을 것 같다는 생각도 드네요

Reply:
```tsx
//변경 전
 const isLgMobile = useMediaQuery('screen and (max-width: 960px)');
 const isXsMobile = useMediaQuery('screen and (max-width: 380px)');

 const itemsPerPage = isXsMobile ? 1 : isLgMobile ? 2 : 3;

//변경 후
const size = useScreenMediaQuery()

const itemsPerPage = ITEMS_PER_PAGE[size];

const ITEMS_PER_PAGE = {
  'lg' : 3 
}
```
사용하는 곳은 이렇게 될 것 같아요

Reply:
윈도우 사이즈가 변경될 때 마다 상태를 변경하고 `window.innerWidth` 접근으로 리플로우를 발생시켜서 성능 상 `window.matchMedia`를 사용하는 것이 좋다고 판단하여 이 코드를 유지하도록 하겠습니다!

다만 `Object.fromEntries`, `Object.entries` 같은 `Object` 이터레이션은 코드를 다소 보기 어렵게 하는 것 같아 올려주신 것과 같이 하드코딩으로 바꾸어 보려고 합니다 😁 (breakpoint 특성 상 변경될 일이 없다시피 하기도 하구요!)

```ts
  const mediaQueries = useMemo<Record<`is${Capitalize<ScreenBreakpoint>}`, MediaQueryList>>(
    () => ({
      isXs: window.matchMedia(`(max-width: ${SCREEN_BREAKPOINT.xs}px)`),
      isSm: window.matchMedia(`(max-width: ${SCREEN_BREAKPOINT.sm}px)`),
      isMd: window.matchMedia(`(max-width: ${SCREEN_BREAKPOINT.md}px)`),
      isLg: window.matchMedia(`(max-width: ${SCREEN_BREAKPOINT.lg}px)`),
      isXl: window.matchMedia(`(max-width: ${SCREEN_BREAKPOINT.xl}px)`),
    }),
    []
  );
```

Reply:
음 윈도우 사이즈가 변경될 때 마다는 아니고 breakpoint가 바뀔 때 마다 상태가 변경되도록 했는데 성능 이슈가 크게 있을까요?

Reply:
앗 그렇네요! 그렇지만 1) 윈도우 사이즈가 변경될 때 마다 event listener를 invoke한다는 점 2) `window.innerWidth` 는 reflow를 발생시킨다는 점에 `window.matchMedia`를 채택하고 싶습니다. `matchMedia`는 주어진 조건에서만 event listener가 invoke되고 reflow를 발생시키지 않는 점이 장점인 것 같습니다.

Reply:
1. window.innerWidth를 읽는게 reflow를 발생시키나요?
2. 매 resize마다 handler를 invoke 시키는게 성능 상 저하요소가 큰가요? resize보다 invoke가 잦은 scroll에도 핸들러 달아서 사용하는 경우가 많은데, 그 보다 자주 발생하는 상황도 아닌 resize에서 핸들러가 invoke 되는 것 때문에 성능저하를 걱정하는게 납득이 잘 안돼요... window.matchMedia과 비교하여 얼마나 성능 차이가 있는지 궁금하고, 그게 크지 않다면 당연히 가독성을 챙기는게 우선이라는 생각입니다.

Reply:
1. `window.innerWidth`, `window.innerHeight` 등 사이즈를 얻어오는 것이 [reflow를 발생한다고 **알려져 있으나**](https://gist.github.com/paulirish/5d52fb081b3570c81e3a) 사실 여부를 확인할 순 없었습니다. 브라우저에서 reflow가 발생했는지 확인할 방법이 있는지 모르겠네요...
2. 성능 저하 요소의 관점보다는 저희가 필요한 시점에만 호출을 받는 것이 추가적인 분기 처리가 없어지고 필요한 코드만 작성할 수 있을 것이라고 생각했습니다.

가독성은 기존의 `Object.fromEntries` 가 너무 장황하게 작성되어서 https://github.com/woowacourse/prolog/pull/1335/commits/fd3c7ebed367798cf37dcb778fcfac938d6c62c8#diff-90ac71c47b7844e8f15f849a684689d43486ae0449794d1a849681e6c58d94afR22-R26 이렇게 바꾸어보았는데 어떨까요,,.. 😓

`window.matchMedia`를 통해 저희가 원하는 것(미디어 쿼리)을 선언적으로 사용할 수 있다고 생각하는데 혹시 다른 단점이 있을지 궁금합니다

Reply:
> window.innerWidth, window.innerHeight 등 사이즈를 얻어오는 것이 [reflow를 발생한다고 알려져 있으나](https://gist.github.com/paulirish/5d52fb081b3570c81e3a) 사실 여부를 확인할 순 없었습니다. 브라우저에서 reflow가 발생했는지 확인할 방법이 있는지 모르겠네요...

아 추가로 reflow 발생 여부는 크롬 개발자 도구 => 성능(performance)에서 확인하실 수 있어요. 제가 테스트 해봤을 때는 브라우저 사이즈를 줄이는 상황 자체가 reflow라서 특별히 다른 reflow는 발견이 안됐어요. 이 부분은 레벨 4에서 배우시게 될 겁니다!

Reply:
혹시 객체 대신 `Map`으로 관리하는 건 어떻게 생각하시나요? 

1) `Object.fromEntries`, `Object.entries` 코드가 반복되지 않음
2) 타입 추론이 잘 됨
3) 하드코딩하지 않아도 가독성이 좋은 것 같음

위와 같은 이유 때문에요!

```ts
type ScreenKey = `is${Capitalize<ScreenBreakpoint>}`;
type Screen = Map<ScreenKey, MediaQueryList>;

const useScreenMediaQuery = () => {
    const mediaQueries = useMemo<Screen>(() => {
    const addIsPrefix = (name: string) => `is${name[0].toUpperCase() + name.slice(1)}`;

    return new Map(
      Object.entries(SCREEN_BREAKPOINT).map(([name, breakpoint]) => {
        return [addIsPrefix(name) as ScreenKey, matchMedia(`(max-width: ${breakpoint}px)`)];
      })
    );
  }, []);

  const [screen, setScreen] = useState(
    new Map([...mediaQueries].map(([name, mediaQuery]) => [name, mediaQuery.matches]))
  );

  useEffect(() => {
    const cleanHandlers = [...mediaQueries].map(([name, mediaQuery]) => {
      const handleMediaQueryChange = ({ matches }: MediaQueryListEvent) => {
        setScreen((screen) => new Map([...screen]).set(name, matches));
      };

      mediaQuery.addEventListener('change', handleMediaQueryChange);

      return () => mediaQuery.removeEventListener('change', handleMediaQueryChange);
    });

    return () => cleanHandlers.forEach((cleanHandler) => cleanHandler());
  }, []);

  return Object.fromEntries(screen);
};

export default useScreenMediaQuery;

```